### PR TITLE
charon: Use new "sources" handlers for getDataset, getNarrative, and getSourceInfo requests

### LINKED ISSUE #150: Charon does not support conditional requests
**Current Behavior**  
When requesting the getDataset? endpoint from a browser with cache, with the headers loaded `If-None-Match` and `If-Modified-Since` both loaded, the CloudFront server responds with a 200 (and possibly connects to origin to grab the data), the proceed to return the same ETag and `Last-Modified-Since` is equal to `If-Modified-Since` in the request.
This may induce additional loading times and significant additional strain on server if the requests are being routed to the origin.

**Expected behavior**  
The server should return 304

**How to reproduce**  
*From browser*:
Open network tab
Open https://nextstrain.org/ncov/global
Refresh repeatedly and observe the network request being carried out fully every time (no disk cache or memory cache are used)

*From command line*:
Run
```
curl 'https://nextstrain.org/charon/getDataset?prefix=/ncov/global' -H 'If-None-Match: "15dae69c5485864e77e2eca2bcc1d139"' -H 'If-Modified-Since: Thu, 16 Apr 2020 03:08:51 GMT' --compressed -v > dump
```
Substitute values of `If-None-Match` and `If-Modified-Since` for the most recent `ETag` and `Last-Modified` as obtained by response headers.
Example output:
```
> If-None-Match: "15dae69c5485864e77e2eca2bcc1d139"
> If-Modified-Since: Thu, 16 Apr 2020 03:08:51 GMT
>
{ [5 bytes data]
< HTTP/1.1 200 OK
< Connection: keep-alive
< X-Powered-By: Express
< Content-Type: application/json
< Content-Length: 1191430
< Date: Thu, 16 Apr 2020 03:09:32 GMT
< Content-Encoding: gzip
< Last-Modified: Thu, 16 Apr 2020 03:08:51 GMT
< X-Amz-Version-Id: dH9GQzvoJifVDaYofZIoU5C74rnIKlTE
< Etag: "15dae69c5485864e77e2eca2bcc1d139"
```
both the ETag and date match, but 200 is returned (304 would be expected)

**Possible solution**  
Possibly related to https://stackoverflow.com/questions/23222430/amazon-cloudfront-not-consistently-returning-304-not-modified-for-unchanged-st

**Your environment: if browsing Nextstrain online**  
 - Operating system:  Linux, Windows
 - User agents:  Chrome, cUrl

**Additional context**  
This may induce additional loading times and significant additional strain on server.

### PR DESCRIPTION
**Pull request by tsibley**
### Description of proposed changes
Several related changes, all detailed in commit messages, but the top-line change is a consolidation of code paths between our Charon endpoints' handlers and the new "sources" handlers powering the RESTful API endpoints.

Resolves #150.

### Testing
- [x] CI passes locally
- [x] CI passes on GitHub (wtf)
- [x] Poking around locally doesn't find any issues

### CODE REVIEW

**Comment by tsibley** [comment]
Tests are passing locally but not on GitHub Actions CI so I'm trying to address that.

**Comment by tsibley** [comment]
Rebased onto latest `master` branch and repushed.

File: src/endpoints/sources.js [hunk 1]
```diff
@@ -620,8 +624,23 @@ function copyHeaders(headerSource, headerNames) {
  * @returns {String} Path for {@link Source#dataset} or {@link Source#narrative}
  */
 
+/* Confused about the duplication below?  It's the documented way to handle
+ * overloaded (e.g. arity-dependent) function signatures.¹  Note that it relies
+ * on the "nestled" or "cuddled" end and start comment markers.
+ *   -trs, 16 June 2022
+ *
+ * ¹ https://github.com/jsdoc/jsdoc/issues/1017
+ */
```
Comment (joverlee521) [inline src/endpoints/sources.js:633]: I'm going to take my time to digest this 🤔

File: src/app.js [hunk 2]
```diff
@@ -130,13 +130,22 @@ app.routeAsync("/charon/getAvailable")
   .getAsync(endpoints.charon.getAvailable);
 
 app.routeAsync("/charon/getDataset")
-  .getAsync(endpoints.charon.getDataset);
+  .getAsync(
+    endpoints.charon.setSourceFromPrefix,
+    endpoints.charon.setDatasetFromPrefix,
+    endpoints.charon.canonicalizeDatasetPrefix,
+    endpoints.charon.getDataset);
```
Comment (joverlee521) [inline src/app.js:137]: It's nice to see how simpler things are after the prefix munging is split out into separate middleware.

File: src/endpoints/sources.js (hunk 1 above)
Comment (tsibley) [reply src/endpoints/sources.js:633]: lmk if you need any digestifs.

File: src/app.js (hunk 2 above)
Comment (tsibley) [reply src/app.js:137]: Yeah! Much of the complexity is from that proto-router.

**Comment by tsibley** [comment]
[CI on the merge commit failed](https://github.com/nextstrain/nextstrain.org/runs/7223516495?check_suite_focus=true#step:5:5) with an obscure internal npm error:

image

which prevented deploy to canary.

Will retry…

**Comment by tsibley** [comment]
…worked when I re-ran the CI run. ¯\_(ツ)_/¯

**Pull request merged**